File: backend/GameApp.Domain/Entities/Scenes/ItemScene.cs
using GameApp.Domain.Entities;
using GameApp.Domain.Enumerates;
using GameApp.Domain.ValueObjects.Items;

namespace GameApp.Domain.ValueObjects.Scenes;

// Class for a scene where you get an item
public class ItemScene : Scene
{

    private readonly Item RewardItem;


    // Constructor
    public ItemScene(SceneName name, SceneDescription description, Biome biome, Item item)
    : base(name, description, biome)
    {
        RewardItem = item;
    }

    // Restore Constructor
    public ItemScene(Guid id, SceneName name, SceneDescription description, Biome biome, Item item)
    : base(id, name, description, biome)
    {
        RewardItem = item;
    }

    // Getter
    public Item GetRewardItem() => RewardItem;

    // Setters 
    public ItemScene SetSceneName(SceneName newName) =>
    new ItemScene(newName, GetDescription(), GetBiome(), RewardItem);

    public ItemScene SetSceneDescription(SceneDescription newDescription) =>
    new ItemScene(GetName(), newDescription, GetBiome(), RewardItem);

    public ItemScene SetBiome(Biome newBiome) =>
    new ItemScene(GetName(), GetDescription(), newBiome, RewardItem);

    public ItemScene SetItem(Item newItem) =>
    new ItemScene(GetName(), GetDescription(), GetBiome(), newItem);

    // To string
    public override string ToString()
    {
        return $"{GetName()} Item Scene({GetGuid()}): " +
               $"Description={GetDescription()}, Biome={GetBiome()}, " +
               $"Reward Item={RewardItem}";
    }
}

File: frontend/gameapp-frontend/src/domain/entities/scenes/item-scene.ts
import { Scene } from "./scene";
import { SceneName } from "../../value-objects/scenes/scene-name";
import { SceneDescription } from "../../value-objects/scenes/scene-description";
import { Biome } from "../../enumerates/biome";
import { Item } from "../items/item";

export class ItemScene extends Scene {
    protected _rewardItem: Item;

    constructor(
        name: SceneName,
        description: SceneDescription,
        biome: Biome,
        item: Item,
        id?: string
    ) {
        super(name, description, biome, id);
        this._rewardItem = item;
    }

    get rewardItem(): Item { return this._rewardItem; }

    setSceneName(newName: SceneName) {
        return new ItemScene(newName, this._description, this._biome, this._rewardItem, this._id);
    }
    setSceneDescription(newDescription: SceneDescription) {
        return new ItemScene(this._name, newDescription, this._biome, this._rewardItem, this._id);
    }
    setBiome(newBiome: Biome) {
        return new ItemScene(this._name, this._description, newBiome, this._rewardItem, this._id);
    }
    setItem(newItem: Item) {
        return new ItemScene(this._name, this._description, this._biome, newItem, this._id);
    }

    toString(): string {
        return `${this._name} Item Scene(${this._id}): Description=${this._description}, Biome=${this._biome}, Reward Item=${this._rewardItem}`;
    }
}
